### ISSUE TITLE
Is DictionaryLiteral<T,T> supported in swiftyJSON? 

### ISSUE THREAD

**Issue by sohaeb** [SwiftyJSON/SwiftyJSON#825]
I am trying to use DictionaryLiteral because it preserves the order of the elements. The code compile, but the JSON is never constructed.

>     var testingDictionary = [
>         
>         "Date" : "Cool",
>         "Time" : [
>             "Morning" :  "Cool",
>             "Night" : "Cool"
>             ]
>        ] as DictionaryLiteral<String, Any>
> 
>      let json = JSON(testingDictionary)
> 
>      print(json)

And the output of the print(json):

> JSON(rawString: "", rawNumber: 0, rawNull: <null>, rawArray: [], rawDictionary: [:], rawBool: false, _type: SwiftyJSON.Type.unknown, _error: Optional(Error Domain=SwiftyJSONErrorDomain Code=999 "It is a unsupported type" UserInfo={NSLocalizedDescription=It is a unsupported type}))

**Comment by TheNamesJames** [comment]
TLDR: No, `DictionaryLiteral` is not supported

#### Why?

> _Disclaimer:_ I'm not hugely familiar with the core development of SwiftyJSON - this answer is based on reading the source code and a couple assumptions.

In the current implementation SwiftyJSON, a dictionary's elements are copied, inserting the keys one at a time - inherently losing its order (see [SwiftyJSON.JSON.init(dictionaryLiteral:)](https://github.com/SwiftyJSON/SwiftyJSON/blob/74cda0fa5ae0be109b60a7dad396bcf92a9d5a13/Source/SwiftyJSON.swift#L533) and [SwiftyJSON.JSON.object & SwiftyJSON.unwrap(_:)](https://github.com/SwiftyJSON/SwiftyJSON/blob/74cda0fa5ae0be109b60a7dad396bcf92a9d5a13/Source/SwiftyJSON.swift#L210-L263)).
So even if an `init` _could_ instantiate a `JSON` object, the order would be unstable (either immediately jumbled, or when the dictionary is modified).

Also, the JSON spec tells us that a JSON object (`Dictionary` being the Swift equivalent) is an "unordered set of name/value pairs" so, besides the SwiftyJSON limitation, the JSON spec itself holds no guarantees on the ordering of name/value pairs (see http://stackoverflow.com/a/4920304).

---

#### Side note for your update:
Swift's `DictionaryLiteral` type and `JSON.init(dictionaryLiteral:)` (from the `ExpressibleByDictionaryLiteral` protocol) are not directly related (see how the argument type is different) - perhaps a little misleading.

`JSON.init(dictionaryLiteral:)` expects arguments more like this:

```swift
let json = JSON(dictionaryLiteral: ("Date", "Cool"),
                ("Time", [
                    "Morning": "Cool",
                    "Night": "Cool"
                    ]))

json.rawString(options: []) ?? ""
// {"Date":"Cool","Time":{"Night":"Cool","Morning":"Cool"}}
```

But, really, `ExpressibleByDictionaryLiteral` (along with the other `ExpressibleByXXLiteral` protocols) is aimed at being used like this:

```swift
let json: JSON = [
    "Date" : "Cool",
    "Time" : [
        "Morning" :  "Cool",
        "Night" : "Cool"
    ]
]
```

I hope this helps?
